src/transformers/models/smollm3/modular_smollm3.py

```diff
@@ -0,0 +1,359 @@
+        self.attention_bias = attention_bias
+        self.attention_dropout = attention_dropout
+
+        if no_rope_layers is None:
+            self.no_rope_layers = [
+                int((layer_idx + 1) % no_rope_layer_interval != 0) for layer_idx in range(num_hidden_layers)
+            ]
+        else:
+            self.no_rope_layers = no_rope_layers
+
+        self.no_rope_layer_interval = no_rope_layer_interval
+
+        # Update layer_types based on sliding window and NoPE pattern
+        if layer_types is None:
+            layer_types = []
+            for layer_idx in range(num_hidden_layers):
+                has_rope = self.no_rope_layers[layer_idx]
+                if use_sliding_window and sliding_window is not None and not has_rope:
+                    layer_types.append("sliding_attention")
+                else:
+                    layer_types.append("full_attention")
+
+        self.layer_types = layer_types
+        layer_type_validation(self.layer_types)
+
+        # Validate the correctness of rotary position embeddings parameters
+        # BC: if there is a 'type' field, move it to 'rope_type'.
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+
+
+class SmolLM3Attention(LlamaAttention):
+    def __init__(self, config: SmolLM3Config, layer_idx: int):
+        super().__init__(config, layer_idx)
+
+        self.use_rope = config.no_rope_layers[layer_idx]
+        self.sliding_window = (
+            config.sliding_window
+            if config.use_sliding_window and config.layer_types[layer_idx] == "sliding_attention"
+            else None
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_embeddings: tuple[torch.Tensor, torch.Tensor],
+        attention_mask: Optional[torch.Tensor],
+        past_key_value: Optional[Cache] = None,
+        cache_position: Optional[torch.LongTensor] = None,
+        **kwargs: Unpack[FlashAttentionKwargs],
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor], Optional[tuple[torch.Tensor]]]:
+        input_shape = hidden_states.shape[:-1]
+        hidden_shape = (*input_shape, -1, self.head_dim)
+
+        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+
+        if self.use_rope:
+            cos, sin = position_embeddings
+            query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
+
+        if past_key_value is not None:
+            if self.use_rope:
+                cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
+            else:
+                cache_kwargs = {"cache_position": cache_position}
```

Comment:
```suggestion
            cache_kwargs = {"cache_position": cache_position}
```
cos and sin were used for sink cache, which is deprecated now so not really neede